### PR TITLE
Add persistence for transaction blocker

### PR DESCRIPTION
**Pull request by kjarosh**

### CODE REVIEW

File: modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        val witness = compareAndExchange(null, acquisition)
+        if (witness == acquisition) {
+            logger.info("Re-entering an acquired transaction lock $acquisition")
+            return
+        }
+        if (witness != null) {
+            throw AlreadyLockedException(witness)
+        }
+
+        logger.info("Transaction lock acquired $acquisition")
```
Comment (Davenury) [inline modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt:44]: Do we need this to be info? There's enough logs in std, maybe it's time to think about cleaning up logs and their levels

File: modules/common/src/main/kotlin/com/github/davenury/common/txblocker/TransactionAcquisition.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * @author Kamil Jarosz
+ */
+data class TransactionAcquisition(
+    @JsonProperty("protocol")
+    val protocol: ProtocolName,
+    @JsonProperty("changeId")
+    val changeId: String,
+) {
+    fun serialize(): String {
+        return ObjectMapper().writeValueAsString(this)
```
Comment (Davenury) [inline modules/common/src/main/kotlin/com/github/davenury/common/txblocker/TransactionAcquisition.kt:17]: Here you'll create object mapper every time. As far as I know it's pretty heavy operation and you won't have kotlin modules installed, so data classes might not work.

File: src/main/kotlin/com/github/davenury/ucac/commitment/gpac/GPACProtocol.kt [hunk 3]
```diff
@@ -143,7 +145,7 @@ class GPACProtocolImpl(
 
         if (!this.transaction.decision) {
             try {
-                transactionBlocker.tryToBlock(ProtocolName.GPAC, message.change.id)
+                transactionBlocker.acquireReentrant(TransactionAcquisition(ProtocolName.GPAC, message.change.id))
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/commitment/gpac/GPACProtocol.kt:148]: ctrl + alt + l?

Comment (kjarosh) [reply src/main/kotlin/com/github/davenury/ucac/commitment/gpac/GPACProtocol.kt:148]: reformatted

File: modules/common/src/main/kotlin/com/github/davenury/common/txblocker/TransactionAcquisition.kt (hunk 2 above)
Comment (kjarosh) [reply modules/common/src/main/kotlin/com/github/davenury/common/txblocker/TransactionAcquisition.kt:17]: moved to a field

File: modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt (hunk 1 above)
Comment (kjarosh) [reply modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt:44]: I would argue that info about tx blocker is important, we certainly should review other information logged

File: modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package com.github.davenury.common.txblocker
+
+import com.github.davenury.common.AlreadyLockedException
+import com.github.davenury.common.NotLockedException
+import com.github.davenury.common.persistence.Persistence
+import org.slf4j.LoggerFactory
+
+private const val TX_BLOCKER_KEY = "txblocker"
```
Comment (Davenury) [inline modules/common/src/main/kotlin/com/github/davenury/common/txblocker/PersistentTransactionBlocker.kt:8]: Is this necessary to be outside of this class?

File: src/test/kotlin/com/github/davenury/ucac/gpac/GPACProtocolSpec.kt [hunk 5]
```diff
@@ -68,10 +70,10 @@ class GPACProtocolSpec {
 
     @Test
     fun `should throw NotElectingYou when ballot number is higher than proposed`(): Unit = runBlocking {
-
-        every { transactionBlockerMock.isAcquired() } returns false
-        every { transactionBlockerMock.tryToBlock(ProtocolName.GPAC, change.id) } just Runs
-        every { transactionBlockerMock.releaseBlock() } just Runs
+        every { transactionBlockerMock.getAcquisition() } returns null
```
Comment (Davenury) [inline src/test/kotlin/com/github/davenury/ucac/gpac/GPACProtocolSpec.kt:73]: Those mocks are used pretty often. Maybe implement `DummyTransactionBlocker`?

**Review by Davenury — DISMISSED** [review]

**Review by Rados13 — APPROVED** [review]

**Review by Davenury — APPROVED** [review]

**Review by Rados13 — APPROVED** [review]

**Pull request merged**